jobs/mysql_restore_job.go

```diff
@@ -0,0 +1,400 @@
+}
+
+// Timeout returns job timeout.
+func (j *MySQLRestoreJob) Timeout() time.Duration {
+	return j.timeout
+}
+
+func binariesInstalled() error {
+	if _, err := exec.LookPath(xtrabackupBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xtrabackupBin)
+	}
+
+	if _, err := exec.LookPath(xbcloudBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xbcloudBin)
+	}
+
+	if _, err := exec.LookPath(xbstreamBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xbstreamBin)
+	}
+
+	if _, err := exec.LookPath(qpressBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", qpressBin)
+	}
+
+	return nil
+}
+
+func prepareRestoreCommands(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+) (xbcloud, xbstream *exec.Cmd, _ error) {
+	xbcloudCmd := exec.CommandContext( //nolint:gosec
+		ctx,
+		xbcloudBin,
+		"get",
+		"--storage=s3",
+		"--s3-endpoint="+config.S3Config.Endpoint,
+		"--s3-access-key="+config.S3Config.AccessKey,
+		"--s3-secret-key="+config.S3Config.SecretKey,
+		"--s3-bucket="+config.S3Config.BucketName,
+		"--s3-region="+config.S3Config.BucketRegion,
+		"--parallel=10",
+		backupName,
+	)
+
+	xbcloudStdout, err := xbcloudCmd.StdoutPipe()
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to get xbcloud stdout pipe")
+	}
+
+	xbstreamCmd := exec.CommandContext( //nolint:gosec
+		ctx,
+		xbstreamBin,
+		"restore",
+		"-x",
+		"--directory="+targetDirectory,
+		"--parallel=10",
+		"--decompress",
+	)
+	xbstreamCmd.Stdin = xbcloudStdout
+
+	return xbcloudCmd, xbstreamCmd, nil
+}
+
+// stdout and stderr could be returned even if rerr is not nil
+func restoreMySQLFromS3(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+) (rerr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	xbcloudCmd, xbstreamCmd, err := prepareRestoreCommands(
+		ctx,
+		backupName,
+		config,
+		targetDirectory,
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := xbcloudCmd.Start(); err != nil {
+		return errors.Wrap(err, "xbcloud start failed")
+	}
+	defer func() {
+		err := xbcloudCmd.Wait()
+		if err == nil {
+			return
+		}
+
+		if rerr != nil {
+			rerr = errors.Wrapf(rerr, "xbcloud wait error: %s", err)
+		} else {
+			rerr = errors.Wrap(err, "xbcloud wait failed")
+		}
+	}()
+
+	if err := xbstreamCmd.Start(); err != nil {
+		return errors.Wrap(err, "xbstream start failed")
+	}
+
+	if err := xbstreamCmd.Wait(); err != nil {
+		return errors.Wrap(err, "xbstream wait failed")
+	}
+
+	return nil
+}
+
+func mySQLActive() (bool, error) {
+	// systemctl is-active returns an exit code 0 if service is active, or non-zero otherwise
+	_, err := exec.Command("systemctl", "is-active", "--quiet", mySQLServiceName).Output()
+	if err == nil {
+		return true, nil
+	}
+
+	if errors.As(err, &exec.ExitError{}) {
+		return false, nil
+	}
+
+	return false, err
+}
+
+func waitForMySQL(expectedActiveStatus bool) error {
```

Comment:
Do we need this? I thought `systemctl start/stop` are blocking operations

Reply:
removed